Source/Core/Common/Image.cpp

```diff
@@ -0,0 +1,183 @@
+// Copyright 2016 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include "Common/Image.h"
+
+#include <csetjmp>
+#include <cstdlib>
+#include <cstring>
+#include <memory>
+#include <png.h>
+#include <vector>
+
+#include "Common/CommonTypes.h"
+#include "Common/Logging/Log.h"
+
+namespace Common
+{
+static int MultiplyAlpha(int alpha, int color)
+{
+  const int temp = (alpha * color) + 0x80;
+  return ((temp + (temp >> 8)) >> 8);
+}
+
+static void PremultiplyData(png_structp png, png_row_infop row_info, png_bytep data)
+{
+  for (png_size_t i = 0; i < row_info->rowbytes; i += 4, data += 4)
+  {
+    const png_byte alpha = data[3];
+    u32 w;
+
+    if (alpha == 0)
+    {
+      w = 0;
+    }
+    else
+    {
+      png_byte red = data[0];
+      png_byte green = data[1];
+      png_byte blue = data[2];
+
+      if (alpha != 0xff)
+      {
+        red = MultiplyAlpha(alpha, red);
+        green = MultiplyAlpha(alpha, green);
+        blue = MultiplyAlpha(alpha, blue);
+      }
+      w = (alpha << 24) | (red << 16) | (green << 8) | (blue << 0);
+    }
+
+    std::memcpy(data, &w, sizeof(w));
+  }
+}
+
+struct ReadProgress
+{
+  const u8* current;
+  const u8* end;
+};
+
+static void ReadCallback(png_structp png, png_bytep data, png_size_t size)
+{
+  ReadProgress* progress = static_cast<ReadProgress*>(png_get_io_ptr(png));
+  for (size_t i = 0; i < size; ++i)
+  {
+    if (progress->current >= progress->end)
+      png_error(png, "Read beyond end of file");  // This makes us longjmp back to LoadPNG
+
+    data[i] = *(progress->current);
+    ++(progress->current);
+  }
+}
+
+static void PNGErrorCallback(png_structp png, png_const_charp error_msg)
+{
+  ERROR_LOG(COMMON, "PNG loading error: %s", error_msg);
+  longjmp(png_jmpbuf(png), 1);
+}
+
+#ifdef _MSC_VER
+#pragma warning(push)
+#pragma warning(disable : 4611)
+// VS shows the following warning even though no C++ objects are destroyed in LoadPNG:
+// "warning C4611: interaction between '_setjmp' and C++ object destruction is non-portable"
+#endif
+
+bool LoadPNG(const std::vector<u8>& input, std::vector<u8>* data_out, u32* width_out,
+             u32* height_out)
+{
+  const bool is_png = !png_sig_cmp(input.data(), 0, input.size());
+  if (!is_png)
+    return false;
+
+  png_struct* png =
+      png_create_read_struct(PNG_LIBPNG_VER_STRING, nullptr, PNGErrorCallback, nullptr);
+  if (!png)
+    return false;
+
+  png_info* info = png_create_info_struct(png);
+  if (!info)
+  {
+    png_destroy_read_struct(&png, &info, nullptr);
+    return false;
+  }
+
+  // It would've been nice to use std::vector for this, but using setjmp safely
+  // means that we have to use this manually managed volatile pointer garbage instead.
+  png_byte** volatile row_pointers_volatile = nullptr;
```

Comment:
Why does it need to be volatile? What does the volatility qualifier ensure in this context?

Reply:
http://en.cppreference.com/w/cpp/utility/program/setjmp

> Upon return to the scope of `setjmp`, all accessible objects, floating-point status flags, and other components of the abstract machine have the same values as they had when `std::longjmp` was executed, except for the non-volatile local variables in `setjmp`'s scope, whose values are indeterminate if they have been changed since the setjmp invocation.

Reply:
Ah, good to know.